README.md
nsrllookup
==========

NEWS
None to speak of at the moment.  However, the CHANGELOG may have
some good stuff for you!


src/unix.hpp
/* $Id: unix.hpp 123 2012-02-08 16:43:07Z rjh $
 * Copyright (c) 2012, Robert J. Hansen <rjh@secret-alchemy.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */


#ifndef __UNIX_HPP
#define __UNIX_HPP

#include <sys/errno.h>
#include <netdb.h>
#include <poll.h>
#include <unistd.h>

#ifdef __APPLE__
#define OPERATING_SYSTEM "Mac OS X"
#endif

#ifdef __linux__
#include <cstring>
#include <sys/socket.h>
#define OPERATING_SYSTEM "Linux"
#endif

#ifdef __FreeBSD__
#define OPERATING_SYSTEM "FreeBSD (cool!)"
#include <sys/socket.h>
#include <netinet/in.h>
#endif

class NetworkSocket
{
public:
    NetworkSocket(std::string host, unsigned short int port) :
        sock(-1), buffer("")
    {
        struct sockaddr_in serv_addr;
        struct hostent *server = gethostbyname(host.c_str());

        if (0 == server ||
                0 > (sock = socket(AF_INET, SOCK_STREAM, 0))) {
            perror("");
            throw NetworkError();
        }

        memset(&serv_addr, 0, sizeof(serv_addr));

        serv_addr.sin_family = AF_INET;
        serv_addr.sin_port = htons(port);
        memcpy((void*) &serv_addr.sin_addr.s_addr,
               (void*) server->h_addr,
               server->h_length);

        if (connect(sock,(struct sockaddr *) &serv_addr,sizeof(serv_addr)) < 0) {
            if (ECONNREFUSED == errno) {
                throw ConnectionRefused();
            } else {
                throw NetworkError();
            }
        }
    }

    virtual ~NetworkSocket()
    {
        if (sock > 0)
            shutdown(sock, SHUT_RDWR);
    }

    void write(const std::string& line)
    {
        // This has a wacky edge case if you're sending 2**32 bytes AND
        // have a network error.  Yes, it's a bug.
        if (line.size() != (size_t) send(sock, line.c_str(), line.size(), 0))
            throw NetworkError();
    }

    void write(const char* buf)
    {
        std::string line(buf);
        write(line);
    }

    std::string read_line()
    {
        pollfd fds = { sock, POLLIN, 0 };
        int poll_code(poll(&fds, 1, 750));
        std::string rv("");
        std::string::iterator iter(std::find(buffer.begin(),
            buffer.end(), '\n'));

        while (0 == poll_code)
            poll_code = poll(&fds, 1, 750);
        if (-1 == poll_code)
            throw NetworkError();
        else if (fds.revents & POLLERR || fds.revents & POLLHUP)
            throw NetworkError();
        else if (fds.revents & POLLIN) {
            if (sock == 0 and buffer == "") {
                throw EOFException();
            }
            while (sock != 0 and iter == buffer.end()) {
                std::vector<char> tmpbuf(8192, '\0');
                int status = 0;

                status = recv(sock, (void*) &tmpbuf[0], 8191, 0);
                if (0 == status) {
                    close(sock);
                    sock = 0;
                }
                else if (-1 == status) {
                    throw NetworkError();
                }
                buffer += std::string(&tmpbuf[0]);
                iter = std::find(buffer.begin(), buffer.end(), '\n');
            }

            rv = std::string(buffer.begin(), iter);
            rv.erase(std::remove(rv.begin(), rv.end(), '\r'), rv.end());

            if (iter != buffer.end())
                iter += 1;
            buffer = std::string(iter, buffer.end());
        }
        return rv;
    }

private:
    int sock;
    std::string buffer;
};

#endif


src/common.cc
/* $Id: common.cc 104 2012-01-30 08:07:11Z rjh $
 * Copyright (c) 2012, Robert J. Hansen <rjh@secret-alchemy.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "common.hpp"
#include <vector>
#include <string>
#include <algorithm>

using std::string;
using std::vector;
using std::find;
using std::remove;
using std::ofstream;

extern ofstream* HIT_FILE;
extern ofstream* MISSES_FILE;
extern NetworkSocket* GLOBAL_SOCK;

vector<string>* tokenize(const string& line, const char delim)
{
    string::const_iterator begin = line.begin();
    string::const_iterator end = line.begin();
    vector<string>* rv = new vector<string>();

    while (begin != line.end()) {
        end = find(begin + 1, line.end(), delim);
        rv->push_back(string(begin, end));
        begin = end + (end == line.end() ? 0 : 1);
    }
    for (size_t idx = 0 ; idx < rv->size() ; ++idx) {
        string& line = rv->at(idx);
        line.erase(remove(line.begin(), line.end(), '\r'), line.end());
        line.erase(remove(line.begin(), line.end(), '\n'), line.end());
    }
    rv->erase(remove(rv->begin(), rv->end(), ""), rv->end());
    return rv;
}

/* This abomination comes to you courtesy of the Win32 API. */
void bomb(int code)
{
    if (GLOBAL_SOCK) {
        delete GLOBAL_SOCK;
    }
    if (HIT_FILE) {
        delete HIT_FILE;
    }
    if (MISSES_FILE) {
        delete MISSES_FILE;
    }
#ifdef WINDOWS
    WSACleanup();
    ExitProcess(code);
#else
    exit(code);
#endif
}


src/win32.hpp
/* $Id: win32.hpp 105 2012-01-30 08:08:50Z rjh $
 * Copyright (c) 2012, Robert J. Hansen <rjh@secret-alchemy.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */


#if _WIN32 || __WIN32__ || __WINDOWS__ || _win32
#ifndef __WIN32_HPP
#define __WIN32_HPP
#define WINDOWS

#undef UNICODE

#include <WinSock2.h>
#include <WS2tcpip.h>
#include <windows.h>

#define OPERATING_SYSTEM "Windows"
#define PACKAGE_VERSION "1.2.2"

class NetworkSocket
{
public:
    NetworkSocket(std::string host, unsigned short int port)
        : sock(socket(AF_INET, SOCK_STREAM, 0)),
          buffer(""),
          temporary_pool(HeapAlloc(GetProcessHeap(),
                                   HEAP_ZERO_MEMORY,
                                   8192)),
          connected(false)
    {
        sockaddr_in server;
        sockaddr_in *sockaddr_ipv4 = 0;
        addrinfo* result, *ptr;
        addrinfo hints;

        ZeroMemory(&hints, sizeof(hints));
        hints.ai_socktype = SOCK_STREAM;
        hints.ai_protocol = IPPROTO_TCP;

        if (INVALID_SOCKET == sock) {
            std::cerr << "Error: couldn't create a socket!\n";
            throw NetworkError();
        }

        if (0 != GetAddrInfo(host.c_str(), NULL, &hints, &result)) {
            std::cerr << "Error: DNS failure (couldn't look up server).\n";
            throw NetworkError();
        }


        for(ptr = result; ptr != NULL ; ptr = ptr->ai_next) {
            if (ptr->ai_family == AF_INET) {
                CopyMemory(&server, ptr->ai_addr, sizeof(server));
                break;
            }
        }
        if (NULL == ptr) {
            std::cerr << "Error: DNS failure (server not found).\n";
            throw NetworkError();
        }

        freeaddrinfo(result);

        server.sin_family = AF_INET;
        server.sin_port = htons(port);

        if (SOCKET_ERROR == connect(sock,
                                    (sockaddr*) &server,
                                    sizeof(server))) {
            throw ConnectionRefused();
        }
        connected = true;
    }

    virtual ~NetworkSocket()
    {
        HeapFree(GetProcessHeap(), 0, temporary_pool);
        closesocket(sock);
    }

    void write(const std::string& line)
    {
        if (SOCKET_ERROR == send(sock, line.c_str(), line.size(), 0)) {
            std::cerr << "Error: couldn't send a packet to the server.\n";
            bomb(0);
        }
    }

    void write(const char* buf)
    {
        std::string line(buf);
        write(line);
    }

    std::string read_line()
    {
        std::string::iterator siter = std::find(buffer.begin(),
                                                buffer.end(),
                                                '\n');
        std::string rv("");
        int bytes_read(0);
        fd_set fds;
        timeval tv;
        int sel_call(0);

        tv.tv_sec = 0;
        tv.tv_usec = 500000;
        FD_ZERO(&fds);
        FD_SET(sock, &fds);

        if (! connected && buffer != "") {
            rv = buffer;
            rv.erase(std::remove(rv.begin(), rv.end(), '\n'), rv.end());
            rv.erase(std::remove(rv.begin(), rv.end(), '\r'), rv.end());
            buffer = "";
            return rv;
        }
        if (! connected && buffer == "") {
            throw EOFException();
        }

        while (connected && siter == buffer.end())
        {
            sel_call = select(0, &fds, NULL, &fds, &tv);
            if (0 == sel_call)
                continue;
            if (SOCKET_ERROR == sel_call) {
                connected = false;
                throw NetworkError();
            }

            ZeroMemory(temporary_pool, 8192);

            bytes_read = recv(sock, (char*) temporary_pool, 8192, 0);
            if (SOCKET_ERROR == bytes_read || 0 == bytes_read) {
                connected = false;
                break;
            }
            buffer += std::string((const char*) temporary_pool,
                                  (const char*) temporary_pool + bytes_read);
            siter = std::find(buffer.begin(), buffer.end(), '\n');
        }
        siter = std::find(buffer.begin(), buffer.end(), '\n');
        rv = std::string(buffer.begin(), siter);
        if (siter != buffer.end())
            siter += 1;
        buffer = std::string(siter, buffer.end());
        return rv;
    }

private:
    SOCKET sock;
    std::string buffer;
    void* temporary_pool;
    bool connected;
};

#endif
#endif


src/common.hpp
/* $Id: common.hpp 105 2012-01-30 08:08:50Z rjh $
 * Copyright (c) 2012, Robert J. Hansen <rjh@secret-alchemy.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#ifndef __COMMON_HPP
#define __COMMON_HPP

#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <algorithm>
#include <cstdlib>
#include <exception>
#include <memory>

void bomb(int code);
int query_server_status();

class NetworkError : public std::exception {
public:
    const char* what() const throw() {
        return "network error";
    }
};
class EOFException : public std::exception {
public:
    const char* what() const throw() {
        return "eof exception";
    }
};
class ConnectionRefused : public std::exception {
public:
    const char* what() const throw() {
        return "connection refused";
    }
};

void parse_options(int argc, char** argv);
std::vector<std::string>* tokenize(const std::string& line,
                                   const char delim = ' ');
void query_server(const std::vector<std::string>& buffer);
void end_connection();


#ifdef _WIN32
#include "win32.hpp"
#else
#include "unix.hpp"
#endif

#endif


src/nsrllookup.cc
/* $Id: nsrllookup.cc 123 2012-02-08 16:43:07Z rjh $
 * Copyright (c) 2012, Robert J. Hansen <rjh@secret-alchemy.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */


#include "common.hpp"

using std::vector;
using std::string;
using std::find;
using std::remove;
using std::ofstream;
using std::cin;
using std::auto_ptr;

string SERVER("localhost");
bool SCORE_HITS(false); // score misses
bool REPORT_STATUS(false); // report server status
int PORT(9120);
ofstream* HIT_FILE(0);
ofstream* MISSES_FILE(0);
NetworkSocket* GLOBAL_SOCK(0);

namespace {
bool is_valid(const string& line)
{
    auto_ptr<vector<string> > tokens(tokenize(line));
    size_t token_count(tokens->size());

    if (token_count < 1)
        return false;
    
    string& hash(tokens->at(0));
    size_t hash_length(hash.size());

    if (hash_length != 32 && hash_length != 40 && hash_length != 64)
        return false;

    string::iterator siter(hash.begin());

    for (siter = hash.begin() ; siter != hash.end() ; ++siter)
        if (! ((*siter >= '0' && *siter <= '9') ||
            (*siter >= 'A' && *siter <= 'F') ||
            (*siter >= 'a' && *siter <= 'f')))
            return false;

    return true;
}
}


int main(int argc, char* argv[])
{
#ifdef WINDOWS
    WSAData wsad;
    if (0 != WSAStartup(MAKEWORD(2, 0), &wsad)) {
        std::cerr <<
                  "Error: could not initialize Winsock.\n\n"
                  "You're running a very old version of Windows.  nsrllookup won't work\n"
                  "on this system.\n";
        bomb(-1);
    }
#endif

    vector<string> buffer;

    parse_options(argc, argv);

    if (REPORT_STATUS) {
        bomb(query_server_status());
    }
    
    try {
        vector<char> buf(4096, '\0');
        while (cin) {
#ifdef WINDOWS
            ZeroMemory(&buf[0], 4096);
#else
            memset(&buf[0], 0, 4096);
#endif
            cin.getline(&buf[0], 4096);
            buf.erase(remove(buf.begin(), buf.end(), '\r'), buf.end());
            buf.erase(remove(buf.begin(), buf.end(), '\n'), buf.end());

            string line(&buf[0]);

            if (is_valid(line)) {
                buffer.push_back(line);
                if (buffer.size() >= 4096) {
                    query_server(buffer);
                    buffer.clear();
                }
            }
        }
    } catch (EOFException&) {
        // pass: this is entirely expected.  Uh, well, maybe.  It should
        // actually be removed, I think...
    }

    if (REPORT_STATUS || 0 < buffer.size()) {
        query_server(buffer);
        buffer.clear();
    }
    end_connection();
    
    bomb(0);
    return 0;
}


src/parse_options.cc
/* $Id: parse_options.cc 104 2012-01-30 08:07:11Z rjh $
 * Copyright (c) 2012, Robert J. Hansen <rjh@secret-alchemy.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */


#include "common.hpp"

using std::string;
using std::ofstream;

extern string SERVER;
extern bool SCORE_HITS;
extern bool REPORT_STATUS;
extern int PORT;
extern ofstream* HIT_FILE;
extern ofstream* MISSES_FILE;

namespace {
string PORTSTRING("9120");

void show_version()
{
    std::cerr <<
              "nsrllookup for " OPERATING_SYSTEM " version " PACKAGE_VERSION "\n\n"
              "The latest version can be found at:\n"
      "        http://rjhansen.github.com/nsrllookup\n\n";
    bomb(0);
}

void show_help()
{
    std::cerr <<
              "nsrllookup for " OPERATING_SYSTEM " version " PACKAGE_VERSION "\n"
              "\n"
              "nsrllookup [-hvukxS] [-U FILE] [-K FILE] [-s SERVER] [-p PORT]\n"
              "\n"
              "        -h: display this help message\n"
              "        -v: display version information\n"
              "        -u: show only unknown hashes (default)\n"
              "        -k: show only known hashes\n"
              "        -S: display server status (if possible)\n"
              "        -U FILE: write unknown hashes to FILE\n"
              "        -K FILE: write known hashes to FILE\n"
              "        -s SERVER: connect to a specified nsrlquery server\n"
              "        -p PORT: connect on a specified port\n"
              "\n"
#ifndef WINDOWS
              "Alternately, you may want to read the manpage: 'man nsrllookup'\n"
              "may have useful information.\n\n"
#endif
              "The latest version can be found at:\n"
              "        http://rjhansen.github.com/nsrllookup\n\n";
    bomb(0);
}

void validate_port()
{
    PORT = 0;

    for (string::const_iterator iter = PORTSTRING.begin() ;
            iter != PORTSTRING.end() ; ++iter) {
        if (*iter < '0' || *iter > '9') {
            std::cerr << "invalid port: must be in range 0 - 65535";
            bomb(0);
        }
        PORT = (PORT * 10) + (*iter - '0');
    }
    if (PORT <= 0 || PORT > 65535) {
        std::cerr << "invalid port: must be in range 0 - 65535";
        bomb(0);
    }
}
}

void parse_options(int argc, char** argv)
{
    int kucount = 0;

    for (int i = 1 ; i < argc ; ++i) {
        string option(argv[i]);
        if (option.size() < 2 || '-' != option[0] || "-h" == option) {
            show_help();
        }
        else if ("-S" == option) {
            REPORT_STATUS = true;
        }
        else if ("-k" == option) {
            SCORE_HITS = true;
            kucount |= 1;
        }
        else if ("-u" == option) {
            SCORE_HITS = false;
            kucount |= 2;
        }
        else if ("-K" == option) {
            if (argc - 1 == i) {
                show_help();
            }
            HIT_FILE = new ofstream(argv[i+1]);
            i += 1;
        }
        else if ("-U" == option) {
            if (argc - 1 == i) {
                show_help();
            }
            MISSES_FILE = new ofstream(argv[i+1]);
            i += 1;
        }
        else if ("-s" == option) {
            if (argc - 1 == i) {
                show_help();
            }
            SERVER = string(argv[i+1]);
            i += 1;
        }
        else if ("-p" == option) {
            if (argc - 1 == i) {
                show_help();
            }
            PORTSTRING = string(argv[i+1]);
            i += 1;
        }
        else if ("-v" == option) {
            show_version();
        }
        else {
            show_help();
        }
    }
    if (3 == kucount) {
        std::cerr <<
                  "Error: the -k and -u flags cannot be used simultaneously (it would be\n"
                  "the same as printing the input file to standard output).  If you need\n"
                  "both known and unknown hashes parsed out in a single pass, use the\n"
                  "-K and -U flags together to print them into separate files.\n"
                  "\n"
                  "Run this program again with the '-h' flag to see usage options.\n";
        bomb(0);
    }

    validate_port();

}


src/query_server.cc
/* $Id: query_server.cc 123 2012-02-08 16:43:07Z rjh $
 * Copyright (c) 2012, Robert J. Hansen <rjh@secret-alchemy.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */


#include "common.hpp"

using std::string;
using std::vector;
using std::auto_ptr;
using std::ofstream;

extern string SERVER;
extern bool SCORE_HITS;
extern int PORT;
extern ofstream* HIT_FILE;
extern ofstream* MISSES_FILE;
extern NetworkSocket* GLOBAL_SOCK;

typedef vector<string> VS;

namespace {

bool supports_v2(true);

class BadHandshake : public std::exception {
public:
    const char* what() const throw() {
        return "bad handshake";
    }
};

class BadQuery : public std::exception {
public:
    const char* what() const throw() {
        return "bad query";
    }
};

class MismatchedResultSet : public std::exception {
public:
    const char* what() const throw() {
        return "mismatched result set";
    }
};

void write_output(const vector<string>& buffer, const string& result_line)
{
    auto_ptr<VS> tokens(tokenize(result_line));
    if (tokens->empty())
        return;

    if (tokens->at(0) != "OK") {
        throw BadQuery();
    }
    const string& results = tokens->at(1);

    if (buffer.size() != results.size())
        throw MismatchedResultSet();

    for (size_t idx = 0 ; idx < buffer.size() ; ++idx) {
        bool hit = results.at(idx) == '1' ? true : false;

        if (0 == MISSES_FILE &&
                0 == HIT_FILE &&
                hit == SCORE_HITS) {
            std::cout << buffer.at(idx) << "\n";
        }
        if (hit && HIT_FILE) {
            (*HIT_FILE) << buffer.at(idx) << "\n";
        }
        if ((! hit) && MISSES_FILE) {
            (*MISSES_FILE) << buffer.at(idx) << "\n";
        }
    }
}

void handshake(NetworkSocket* sock, string version)
{
    version = "Version: " + version + "\r\n";
    sock->write(version.c_str());
    auto_ptr<VS> tokens(tokenize(sock->read_line()));
    if (tokens->size() == 0 || tokens->at(0) != "OK") {
        throw BadHandshake();
    }
}

string make_query(const vector<string>& buffer)
{
    string rv = "query";

    if (buffer.size() == 0) {
        throw BadQuery();
    }

    for (size_t idx = 0 ; idx < buffer.size() ; ++idx) {
        auto_ptr<VS> tokens(tokenize(buffer.at(idx)));
        rv += " " + tokens->at(0);
    }

    rv += "\r\n";

    return rv;
}


void query_handler(const vector<string>& buffer)
{
    if (buffer.empty() || 0 == GLOBAL_SOCK)
        return;

    try {
        GLOBAL_SOCK->write(make_query(buffer));
        string line = GLOBAL_SOCK->read_line();
        write_output(buffer, line);
    }
    catch (BadQuery&) {
        std::cerr << "Error: server didn't like our query.\n";
        bomb(-3);
    }
    catch (NetworkError&) {
        std::cerr << "Error: network failure.\n";
        bomb(-4);
    }
    catch (MismatchedResultSet&) {
        std::cerr << "Error: mismatched result set.\n";
        bomb(-5);
    }
    catch (std::exception&) {
        std::cerr << "Error: unknown error (WTF?)\n";
        bomb(-6);
    }
}
}

void end_connection()
{
    try {
        if (0 != GLOBAL_SOCK) {
            GLOBAL_SOCK->write("BYE\r\n");
        }
    }
    catch (std::exception&) {
        // pass: we're closing the connection anyway
    }
}

void query_server(const vector<string>& buffer)
{
    try {
        if (0 == GLOBAL_SOCK) {
            GLOBAL_SOCK = new NetworkSocket(SERVER, PORT);
            handshake(GLOBAL_SOCK, "2.0");
        }
        query_handler(buffer);
    } catch (ConnectionRefused&) {
        std::cerr << "Error: connection refused\n";
        bomb(-1);
    } catch (BadHandshake&) {
        std::cerr << "Error: server handshake failed\n";
        bomb(-2);
    }
}

int query_server_status()
{
    int rv(0);
    try {
        if (0 == GLOBAL_SOCK) {
            GLOBAL_SOCK = new NetworkSocket(SERVER, PORT);
            handshake(GLOBAL_SOCK, "2.0");
        }
        GLOBAL_SOCK->write("STATUS\r\n");
        auto_ptr<VS> tokens(tokenize(GLOBAL_SOCK->read_line()));

        if (tokens->size() && tokens->at(0) == "NOT") {
            std::cerr << "Server does not support status queries.\n";
            rv = -7;
        } else if (tokens->size() < 8) {
            std::cerr << "Server returned an incorrect status string\n";
            rv = -8;
        } else {
            long num_of_hashes(atol(tokens->at(1).c_str()));
            std::string bitlength("???");
            double load[3] = {0.0, 0.0, 0.0};

            if (tokens->at(2) == "MD5") {
                bitlength = "128";
            } else if (tokens->at(2) == "SHA-1") {
                bitlength = "160";
            } else if (tokens->at(2) == "SHA-256") {
                bitlength = "256";
            }

            load[0] = atof(tokens->at(5).c_str());
            load[1] = atof(tokens->at(6).c_str());
            load[2] = atof(tokens->at(7).c_str());

            std::streamsize prec = std::cout.precision();
            std::cout.precision(3);
            std::cout << "Server reports " << num_of_hashes << " " 
                << bitlength << "-bit hashes, load avgs "
                << load[0] << " " << load[1] << " " << load[2]
                << "\n";
            std::cout.precision(prec);
        }
    } catch (ConnectionRefused&) {
        std::cerr << "Error: connection refused\n";
        rv = -1;
    } catch (BadHandshake&) {
        std::cerr << "Error: server handshake failed\n";
        rv = -2;
    }
    return rv;
}

